### ISSUE TITLE
500 error when updating dotfiles (files starting with a dot)

### ISSUE THREAD

**Issue by Neamar** [nextcloud/server#21209]
### Steps to reproduce
1. Mount an external storage through `mount.cifs`: `mount.cifs -o seal,user=USER,pass=PASSWORD,uid=www-data,gid=www-data,dir_mode=0770 //somewhere.internet /mnt/storage --verbose`
2. Using the nextcloud docker image, set up a volume between `/mnt/storage/nextcloud/data:/var/www/html/data`
3. Try uploading a file starting with a dot. This will work.
4. Try uploading a new version of the same file. You now get a 500 Internal Server Error

### Expected behaviour
This should work reliably, and the dotfile should be synced.

### Actual behaviour
Error 500. Logs say "Error while copying file to target location (copied bytes: 0, expected filesize: 2361 )".

### Server configuration

**Operating system:** Ubuntu 18.04

**Web server:** Nginx + nextcloud docker image with apache

**Database:** PG

**PHP version:**  7.3.18

**Nextcloud version:** (see Nextcloud admin page) 18.0.4

**Updated from an older Nextcloud/ownCloud or fresh install:** fresh install

**Where did you install Nextcloud from:** `FROM nextcloud:18-apache`

**Signing status:**
<details>
<summary>Signing status</summary>

```
Login as admin user into your Nextcloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results here.

No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
If you have access to your command line run e.g.:
sudo -u www-data php occ app:list
from within your Nextcloud installation folder
```
</details>

**Are you using external storage, if yes which one:** no, not at nextcloud level -- but I'm using cifs at the OS level.

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:**

**Operating system:**

#### Web server error log
<details>
<summary>Web server error log</summary>

```
N/A, just "500"
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
```
{"reqId":"wDYPFFx6ZfUpObqhifo0","level":3,"time":"2020-06-03T13:22:26+00:00","remoteAddr":"172.17.0.1","user":"neamar","app":"no app in context","method":"PUT","url":"/remote.php/dav/files/neamar/svg/back
ups/.zshrc","message":{"Exception":"Sabre\\DAV\\Exception","Message":"Error while copying file to target location (copied bytes: 0, expected filesize: 2361 )","Code":0,"Trace":[{"file":"/var/www/html/3rdp
arty/sabre/dav/lib/DAV/Server.php","line":1130,"function":"put","class":"OCA\\DAV\\Connector\\Sabre\\File","type":"->","args":["*** sensitive parameter replaced ***"]},{"file":"/var/www/html/3rdparty/sabr
e/dav/lib/DAV/CorePlugin.php","line":513,"function":"updateFile","class":"Sabre\\DAV\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"function":"httpPut","class":"Sabre\\DAV\\CoreP
lugin","type":"->","args":[{"absoluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/
var/www/html/3rdparty/sabre/event/lib/EventEmitterTrait.php","line":105,"function":"call_user_func_array","args":[[{"__class__":"Sabre\\DAV\\CorePlugin"},"httpPut"],[{"absoluteUrl":"http://nextcloud.neama
r.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":479,"
function":"emit","class":"Sabre\\Event\\EventEmitter","type":"->","args":["method:PUT",[{"absoluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\
\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":254,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"ab
soluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/html/apps/dav/lib/Serve
r.php","line":319,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type"
:"->","args":[]},{"file":"/var/www/html/remote.php","line":165,"args":["/var/www/html/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/var/www/html/apps/dav/lib/Connector/Sabre/File.p
hp","Line":219,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Linux) mirall/2.6.4git (Nextcloud)","version":"18.0.4.2"}
{"reqId":"wDYPFFx6ZfUpObqhifo0","level":4,"time":"2020-06-03T13:22:26+00:00","remoteAddr":"172.17.0.1","user":"neamar","app":"webdav","method":"PUT","url":"/remote.php/dav/files/neamar/svg/backups/.zshrc"
,"message":{"Exception":"Sabre\\DAV\\Exception","Message":"Error while copying file to target location (copied bytes: 0, expected filesize: 2361 )","Code":0,"Trace":[{"file":"/var/www/html/3rdparty/sabre/
dav/lib/DAV/Server.php","line":1130,"function":"put","class":"OCA\\DAV\\Connector\\Sabre\\File","type":"->","args":["*** sensitive parameter replaced ***"]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/D
AV/CorePlugin.php","line":513,"function":"updateFile","class":"Sabre\\DAV\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"function":"httpPut","class":"Sabre\\DAV\\CorePlugin","typ
e":"->","args":[{"absoluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/htm
l/3rdparty/sabre/event/lib/EventEmitterTrait.php","line":105,"function":"call_user_func_array","args":[[{"__class__":"Sabre\\DAV\\CorePlugin"},"httpPut"],[{"absoluteUrl":"http://nextcloud.neamar.fr/remote
.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":479,"function":"
emit","class":"Sabre\\Event\\EventEmitter","type":"->","args":["method:PUT",[{"absoluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},
{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":254,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"absoluteUrl":
"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/html/apps/dav/lib/Server.php","lin
e":319,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->","args
":[]},{"file":"/var/www/html/remote.php","line":165,"args":["/var/www/html/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/var/www/html/apps/dav/lib/Connector/Sabre/File.php","Line":
219,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Linux) mirall/2.6.4git (Nextcloud)","version":"18.0.4.2"}
{"reqId":"la3fzvrqFLm9kzOytjnv","level":3,"time":"2020-06-03T13:23:23+00:00","remoteAddr":"172.17.0.1","user":"neamar","app":"PHP","method":"PUT","url":"/remote.php/dav/files/neamar/svg/backups/.zshrc","m
essage":"file_put_contents(/var/www/html/data/neamar/files/svg/backups/.zshrc): failed to open stream: Permission denied at /var/www/html/lib/private/Files/Storage/Local.php#503","userAgent":"Mozilla/5.0
(Linux) mirall/2.6.4git (Nextcloud)","version":"18.0.4.2"}
{"reqId":"la3fzvrqFLm9kzOytjnv","level":3,"time":"2020-06-03T13:23:23+00:00","remoteAddr":"172.17.0.1","user":"neamar","app":"no app in context","method":"PUT","url":"/remote.php/dav/files/neamar/svg/back
ups/.zshrc","message":{"Exception":"Sabre\\DAV\\Exception","Message":"Error while copying file to target location (copied bytes: 0, expected filesize: 2376 )","Code":0,"Trace":[{"file":"/var/www/html/3rdp
arty/sabre/dav/lib/DAV/Server.php","line":1130,"function":"put","class":"OCA\\DAV\\Connector\\Sabre\\File","type":"->","args":["*** sensitive parameter replaced ***"]},{"file":"/var/www/html/3rdparty/sabr
e/dav/lib/DAV/CorePlugin.php","line":513,"function":"updateFile","class":"Sabre\\DAV\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"function":"httpPut","class":"Sabre\\DAV\\CoreP
lugin","type":"->","args":[{"absoluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/
var/www/html/3rdparty/sabre/event/lib/EventEmitterTrait.php","line":105,"function":"call_user_func_array","args":[[{"__class__":"Sabre\\DAV\\CorePlugin"},"httpPut"],[{"absoluteUrl":"http://nextcloud.neama
r.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":479,"
function":"emit","class":"Sabre\\Event\\EventEmitter","type":"->","args":["method:PUT",[{"absoluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\
\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":254,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"ab
soluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/html/apps/dav/lib/Serve
r.php","line":319,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type"
:"->","args":[]},{"file":"/var/www/html/remote.php","line":165,"args":["/var/www/html/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/var/www/html/apps/dav/lib/Connector/Sabre/File.p
hp","Line":219,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Linux) mirall/2.6.4git (Nextcloud)","version":"18.0.4.2"}
{"reqId":"la3fzvrqFLm9kzOytjnv","level":4,"time":"2020-06-03T13:23:23+00:00","remoteAddr":"172.17.0.1","user":"neamar","app":"webdav","method":"PUT","url":"/remote.php/dav/files/neamar/svg/backups/.zshrc"
,"message":{"Exception":"Sabre\\DAV\\Exception","Message":"Error while copying file to target location (copied bytes: 0, expected filesize: 2376 )","Code":0,"Trace":[{"file":"/var/www/html/3rdparty/sabre/
dav/lib/DAV/Server.php","line":1130,"function":"put","class":"OCA\\DAV\\Connector\\Sabre\\File","type":"->","args":["*** sensitive parameter replaced ***"]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/D
AV/CorePlugin.php","line":513,"function":"updateFile","class":"Sabre\\DAV\\Server","type":"->","args":["*** sensitive parameters replaced ***"]},{"function":"httpPut","class":"Sabre\\DAV\\CorePlugin","typ
e":"->","args":[{"absoluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/htm
l/3rdparty/sabre/event/lib/EventEmitterTrait.php","line":105,"function":"call_user_func_array","args":[[{"__class__":"Sabre\\DAV\\CorePlugin"},"httpPut"],[{"absoluteUrl":"http://nextcloud.neamar.fr/remote
.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":479,"function":"
emit","class":"Sabre\\Event\\EventEmitter","type":"->","args":["method:PUT",[{"absoluteUrl":"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},
{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":254,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"absoluteUrl":
"http://nextcloud.neamar.fr/remote.php/dav/files/neamar/svg/backups/.zshrc","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/html/apps/dav/lib/Server.php","lin
e":319,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->","args
":[]},{"file":"/var/www/html/remote.php","line":165,"args":["/var/www/html/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/var/www/html/apps/dav/lib/Connector/Sabre/File.php","Line":
219,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Linux) mirall/2.6.4git (Nextcloud)","version":"18.0.4.2"}
{"reqId":"DWcXB96RI1mstfUxymRk","level":3,"time":"2020-06-03T13:28:28+00:00","remoteAddr":"172.17.0.1","user":"neamar","app":"PHP","method":"PUT","url":"/remote.php/dav/files/neamar/svg/backups/.zshrc","m
essage":"file_put_contents(/var/www/html/data/neamar/files/svg/backups/.zshrc): failed to open stream: Permission denied at /var/www/html/lib/private/Files/Storage/Local.php#503","userAgent":"Mozilla/5.0
(Linux) mirall/2.6.4git (Nextcloud)","version":"18.0.4.2"}
```
```
</details>

This is probably related to https://github.com/nextcloud/server/issues/9663 but I'm not using an external storage volume, it's a plain old mount at the OS level.

Also worth saying: if I docker enter the container, cd to the data folder, I *can* create a dot file that appears on Samba.

**Comment by kesselb** [comment]
> file_put_contents(/var/www/html/data/neamar/files/svg/backups/.zshrc): failed to open stream: Permission denied at /var/www/html/lib/private/Files/Storage/Local.php#503

Check if the user running Nextcloud inside the container is able to write to that directory.

**Comment by Neamar** [comment]
Every other file is working fine :) Directory is owned by www-data (and it seems Samba doesn't really keep file owner anyway, so when mounted I specify www-data as the owner and every single file is written with www-data as the owner).

If I run `docker enter` and then `touch data/.test`, the file is created:

```sh
$ docker enter nextcloud bash
root@392daa5cf0d5:/var/www/html# cd data/
root@392daa5cf0d5:/var/www/html/data# touch .test
root@392daa5cf0d5:/var/www/html/data# ls -al
total 5124
drwxrwx---  2 www-data www-data      0 Jun  4 09:02 .
drwxrwxrwx 14 www-data root       4096 May 23 09:41 ..
-rwxr-xr-x  1 www-data www-data    542 May 16 21:08 .htaccess
-rwxr-xr-x  1 www-data www-data      1 May 22 19:19 .ocdata
-rwxr-xr-x  1 www-data www-data      0 Jun  4 09:02 .test
drwxrwx---  2 www-data www-data      0 May 22 21:58 appdata_ocxsjlj2bm7m
drwxrwx---  2 www-data www-data      0 May 22 09:00 neamar
-rwxr-xr-x  1 www-data www-data 378079 Jun  4 08:12 nextcloud.log
```

I also tried looking into the specific directory where the error is coming from:

```sh
root@392daa5cf0d5:/var/www/html/data/neamar/files/svg/backups# ls -al
total 5120
drwxrwx--- 2 www-data www-data      0 Jun  3 13:27 .
drwxrwx--- 2 www-data www-data      0 May 21 13:07 ..
-rwxr-xr-x 1 www-data www-data    332 May 10  2018 .npmrc
-rwxr-xr-x 1 www-data www-data   2329 Sep 26  2019 .zshrc
drwxrwx--- 2 www-data www-data      0 Jun  3 13:16 Code
-rwxr-xr-x 1 www-data www-data    601 Jun  3 13:22 README.md
-rwxr-xr-x 1 www-data www-data 849664 Jun  4 09:00 backup-ubuntu.tar.gz.enc
-rwxr-xr-x 1 www-data www-data   1903 Jun  3 13:17 settings.json
```

From the docker container, I can create a `.test` file:

```
root@392daa5cf0d5:/var/www/html/data/neamar/files/svg/backups# touch .test
root@392daa5cf0d5:/var/www/html/data/neamar/files/svg/backups# ls -al
total 6144
drwxrwx--- 2 www-data www-data      0 Jun  4 10:21 .
drwxrwx--- 2 www-data www-data      0 May 21 13:07 ..
-rwxr-xr-x 1 www-data www-data    332 May 10  2018 .npmrc
-rwxr-xr-x 1 www-data www-data      0 Jun  4 10:21 .test
-rwxr-xr-x 1 www-data www-data   2329 Sep 26  2019 .zshrc
drwxrwx--- 2 www-data www-data      0 Jun  3 13:16 Code
-rwxr-xr-x 1 www-data www-data    601 Jun  3 13:22 README.md
-rwxr-xr-x 1 www-data www-data 849664 Jun  4 09:00 backup-ubuntu.tar.gz.enc
-rwxr-xr-x 1 www-data www-data   1903 Jun  3 13:17 settings.json
```

And the Nextcloud client will also synchronize a fille called `.test` if I create one. But uploading a new version of `.zshrc` will return a 500 again, and same thing for `.test`.

**tl;dr: creating a dotfile works well, but I can't update it after creation.**

**Comment by szaimen** [comment]
Looks like a duplicate of https://github.com/nextcloud/server/issues/19317 to me
